verification/certifyKey_test.go

```diff
@@ -124,15 +423,210 @@ func testCertifyKey(d TestDPEInstance, t *testing.T) {
 	}
 
+	// Get DPE leaf certificate from CertifyKey
 	certifyKeyResp, err := client.CertifyKey(&certifyKeyReq)
 	if err != nil {
 		t.Fatalf("Could not certify key: %v", err)
 	}
-	checkCertificateStructure(t, certifyKeyResp.Certificate)
+
+	// Get root and intermediate certificates to validate certificate chain of leaf cert
+	getCertificateChainResp, err := client.GetCertificateChain()
+	if err != nil {
+		t.Fatalf("Could not get Certificate Chain: %v", err)
+	}
+
+	leafCertBytes := certifyKeyResp.Certificate
+	certChainBytes := getCertificateChainResp.CertificateChain
+
+	// Run X.509 linter on full certificate chain and file issues for errors
+	leafCert := checkCertificateStructure(t, leafCertBytes)
+	if leafCert == nil {
+		t.Fatalf("Could not obtain leaf certificate from CertifyKey command.")
+	}
+	t.Logf("Leaf certificate is DER encoded")
+
+	certChain := checkCertificateChain(t, certChainBytes)
+	if certChain == nil || len(certChain) == 0 {
+		t.Fatalf("Could not obtain certificate chain from GetCertificateChain command.")
+	}
+	t.Logf("Certificate chain is DER encoded")
+
+	// Validate that all X.509 fields conform with the format defined in the DPE iRoT profile
+	validateCertifyKeyFields(t, leafCert)
+
+	// Validate that proper TcbInfo extensions are populated with expected measurements and values as defined in the DPE iRoT profile
+
+	// Validate correct KeyUsage and BasicConstraints extensions for all DICE certificates
+
+	// Ensure full certificate chain has valid signatures
+	validateCertChain(t, certChain, leafCert)
 
 	// TODO: When DeriveChild is implemented, call it here to add more TCIs and call CertifyKey again.
 }
+
+// Validate signature of certificates in certificate chain passed.
+func validateSignature(t *testing.T, certchain []*x509.Certificate) {
+	t.Helper()
+	foundError := false
+	for i := 0; i < len(certchain)-1; i++ {
+		err := certchain[i].CheckSignatureFrom(certchain[i+1])
+		if err != nil {
+			t.Logf("Signature validation failed for certificate %s in chain with error %s", certchain[i].Subject.String(), err.Error())
+			foundError = true
+		} else {
+			t.Logf("Signature validation passed %s", certchain[i].Subject.String())
+		}
+
+	}
+
+	root := certchain[len(certchain)-1]
+	//err := root.CheckSignatureFrom(root)
+	err := root.CheckSignature(root.SignatureAlgorithm, root.RawTBSCertificate, root.Signature)
+	if err != nil {
+		t.Logf("Root certificate  %s Signature validation failed with error %s", root.Subject.String(), err.Error())
+		foundError = true
+	}
+	t.Logf("Root certificate Signature validation passed %s\n", root.Subject.String())
+
+	if foundError {
+		// TODO: Fail the test when the certificate is expected to pass
+		t.Logf("Found certificates with mismatching signature")
+	}
+}
+
+// Build certificate chain and calls to validateSignature on each chain.
+func validateCertChain(t *testing.T, certChain []*x509.Certificate, leafCert *x509.Certificate) {
+	t.Helper()
+	if leafCert != nil {
```

Comment:
A lot of this code looks to be duplicated. Can we move the if statement to only do the `leafCert` calls if `leafCert != nil`, but otherwise keep the other code common.

Specifically, I think lines 510-518 are the only ones that differ.

Reply:
Removed duplication